### ISSUE TITLE
Performance degradation post commit 465394798ebb7e53bb859295f6e6f8caa1c42044

### ISSUE THREAD

**Issue by gregbarton** [deeplearning4j/deeplearning4j#9833]
#### Issue Description

Running GenerateTxtCharCompGraphModel example: https://github.com/deeplearning4j/deeplearning4j-examples/blob/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/advanced/modelling/charmodelling/generatetext/GenerateTxtCharCompGraphModel.java

All commits post 465394798ebb7e53bb859295f6e6f8caa1c42044 in deeplearning4j master branch experience a steep performance degradation when running this example code in CPU mode.  It seems like CPU is not being fully utilized. 

#### Version Information

* Deeplearning4j master branch, all commits subsequent to 465394798ebb7e53bb859295f6e6f8caa1c42044
* Ubuntu 22.04.1

#### Additional Information

Execution of the example using commit 465394798ebb7e53bb859295f6e6f8caa1c42044 has this level of CPU utilization:

![image](https://user-images.githubusercontent.com/1505300/199163568-06ea4df0-70d7-476b-854c-128fb58c8ae1.png)

Subsequent commits that build have this decreased level of CPU utilization:

![image](https://user-images.githubusercontent.com/1505300/199163639-55bd7434-c4da-4e72-b4cf-c724b6aae447.png)

**Comment by agibsonccc** [comment]
@gregbarton thanks for flagging!  Let me take a look and see what's going on there!

**Comment by agibsonccc** [comment]
@gregbarton I ran this through yourkit and found most of the time in the example being spent in the indexing mechanisms. (Basically the get(INDArrayIndex...) calls. 

I know for a fact I had fixed some bugs here and there but it appears we had a performance regression in the process. I likely won't undo the changes but I'll ensure that we get equivalent performance by profiling this example and making the appropriate changes there. Thanks againf or flagging!
